optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/DroolsConstraint.java

```
@@ -57,13 +60,52 @@ public DroolsConstraint(DroolsConstraintFactory<Solution_> constraintFactory,
-                .build(ruleItemBuilderList.toArray(new RuleItemBuilder<?>[0]));
+        return streams;
+    }
+
+    private Collection<DroolsAbstractConstraintStream<Solution_>> assembleAllStreams(
+            DroolsAbstractConstraintStream<Solution_> parent) {
+        final Set<DroolsAbstractConstraintStream<Solution_>> streams = new LinkedHashSet<>();
+        streams.add(parent);
+        for (DroolsAbstractConstraintStream<Solution_> child: parent.getChildStreams()) {
+            streams.addAll(assembleAllStreams(child));
+        }
+        return streams;
+    }
+
+    /**
+     * Retrieves Drools rules required to process this constraint.
+     *
+     * @param ruleLibrary never null. Cache of rules already generated by previous constraints. This method will use
+     * {@link Map#computeIfAbsent(Object, Function)} to add rules from {@link DroolsAbstractConstraintStream}s which we
+     * have not yet processed. This way, we will only generate one rule per stream, allowing Drools to reuse the
+     * computations.
+     * @param scoreHolderGlobal never null. The Drools global used to track changes to score within rule consequences.
+     * @return never null. List of all Drools {@link Rule}s required by this constraint. May update ruleLibrary as a
+     * side effect.
+     */
+    public List<Rule> createRules(Map<DroolsAbstractConstraintStream<Solution_>, Rule> ruleLibrary,
+            Global<? extends AbstractScoreHolder> scoreHolderGlobal) {
+        Set<DroolsAbstractConstraintStream<Solution_>> streams = assembleAllStreams();
+        List<Rule> rules = new ArrayList<>();
+        streams.forEach(stream -> {
+            Rule rule = ruleLibrary.computeIfAbsent(stream, key ->
+                    key.buildRule(this, scoreHolderGlobal).orElse(null));
+            if (rule != null) {
+                rules.add(rule);
+            }
+        });
+        int expectedRuleCount = rules.size();
+        if (expectedRuleCount == 0) {
```

Should this really fail fast?
I don't think Bavet fails on having an empty Constraint array.
We should have a unit test for that for an empty constraint array.

I'd argue that the ruleList.size() <= streamSet.size()
(and both can be zero)